### DISCUSSION TITLE
Clozure CL 1.13

### DISCUSSION THREAD

**Discussion by xrme** [Clozure/ccl#511] [General]
This is Clozure CL 1.13.

There are two ways to obtain a copy of the release.

If you are not interested in tracking or updating CCL sources at all, you may download an archive file of the form `ccl-1.13.platform.gz`. This archive will contain everything you need to run CCL (including complete source code and binaries), but not a copy of the Git repository.

If, on the other hand, you would like to be able to pull updates from Git, follow these steps:

1. Obtain source code by cloning the repository (with `git clone https://github.com/Clozure/ccl.git`)
2. Download and extract the bootstrapping binaries for your system. Because CCL is written in itself, you need an already-working copy of CCL to compile it. The archive also contains a database derived from the system header files which is used by the FFI.

For example, to get a copy of CCL for Linux/x86, you would say:
```
git clone https://github.com/Clozure/ccl.git ccl
curl -L -O https://github.com/Clozure/ccl/releases/download/v1.13/linuxx86.tar.gz
cd ccl
tar xf ../linuxx86.tar.gz
```

After unpacking the binaries, you can rebuild CCL by evaluating `(rebuild-ccl :full t)` as usual.

Unfortunately, on Windows, `(rebuild-ccl :full t)` will not work. Therefore, you must build the lisp kernel binary (a program written in C and assembly language) separately. To do this, use the following commands:
```
cd lisp-kernel/darwinx8664
# or lisp-kernel/win64 or lisp-kernel/win32
make clean
make
```

When that is done, start the lisp and evaluate `(rebuild-ccl :clean t)`. This does the same thing as `:full t`, except it will not attempt to rebuild the lisp kernel for you.

The name of the lisp kernel binary (which is the executable that you invoke to start CCL) is different for each platform. For instance, on macOS, the name is `dx86cl64`, so you could start CCL by running `./dx86cl64`. See the [manual](https://ccl.clozure.com/docs/ccl.html#installing-and-running-clozure-cl) for more details.

Binaries were built on the following systems:

| Platform | Version |
| --- | --- |
| darwin/x86 | Xcode tools 15.4 on macOS Sonoma 14.6.1 |
| freebsd/x86 | FreeBSD 13.3 |
| linux/arm | Raspberry Pi OS (Legacy) (based on Debian 11 bullseye) on Raspberry Pi |
| linux/x86 | Ubuntu 22.04.4 LTS |
| solaris/x86 | OmniOS LTS (r151046) |
| windows/x86 | Windows 10 with MSYS2 (MINGW32 environment for 32-bit, and MINGW64 environment for 64-bit) |

Both 64-bit and 32-bit versions of the lisp are included on the x86 platforms (except for macOS, which doesn't support running 32-bit x86 binaries).

This release does not contain support for Apple Silicon (ARM64) processors.

For convenience, we recommend that you use a script to invoke CCL. Please see the sample [ccl](https://github.com/Clozure/ccl/blob/c994217adc56b3f8a564526cee4695893ac84d86/scripts/ccl) and [ccl64](https://github.com/Clozure/ccl/blob/c994217adc56b3f8a564526cee4695893ac84d86/scripts/ccl64) scripts in the [scripts](https://github.com/Clozure/ccl/tree/c994217adc56b3f8a564526cee4695893ac84d86/scripts) directory for examples you can customize as needed.

### The interface database
This release uses https://github.com/Clozure/ccl-ffigen to process header files for FreeBSD, Linux, and Solaris-ish systems. See that link for more details, but the short version is that you may run into differences when using the `#_` and `#$` reader macros. Additionally, many fewer header files are processed by default, so if your code accesses a fairly obscure foreign name, you might find that it is now missing. Please report problems by [creating an issue](https://github.com/Clozure/ccl/issues/new).

The Mac and Windows ports carry forward the existing interface databases unchanged. They will be updated in a future release.

### 32-bit x86
A 32-bit x86 Lisp is again provided for FreeBSD and Solaris-ish systems (in addition to a 64-bit x86 version).

**Full Changelog**: https://github.com/Clozure/ccl/compare/v1.12.2...v1.13

<hr /><em>This discussion was created from the release <a href='https://github.com/Clozure/ccl/releases/tag/v1.13'>Clozure CL 1.13</a>.</em>

**Comment by gmpalter** [comment]
`(rebuild-ccl :full t)` is working on Windows again. (I just did it on Win10 64-bit.)

**Reply by xrme** [reply]
I didn't think that ever worked before because Windows wouldn't delete an open file.
